esmvaltool/preprocessor/_derive/sm.py

```diff
@@ -0,0 +1,41 @@
+"""Derivation of variable `sm`."""
+
+import cf_units
+from iris import Constraint
+import numpy as np
+
+from ._derived_variable_base import DerivedVariableBase
+
+
+class DerivedVariable(DerivedVariableBase):
+    """Derivation of variable `sm`."""
+
+    # Required variables
+    _required_variables = {
+        'vars': [{
+            'short_name': 'mrsos',
+            'field': 'T2{frequency}s'
+        }]
+    }
+
+    def calculate(self, cubes):
+        """Compute soil moisture.
+
+        Note
+        ----
+        Convert moisture content of soil layer (kg/m2) into volumetric soil
+        moisture (m3/m3), assuming density of water 998.2 kg/m2 (at temperature
+        20 deg C).
+
+        """
+        mrsos_cube = cubes.extract_strict(
+            Constraint(name='moisture_content_of_soil_layer'))
+
+        depth = mrsos_cube.coord('depth').bounds
+        layer_thickness = depth[..., 1] - depth[..., 0]
+
+        sm_cube = mrsos_cube / layer_thickness / 998.2
+        sm_cube.units = cf_units.Unit('m3 m^-3')
+        sm_cube.data = np.ma.array(sm_cube.data, dtype=np.dtype('float32'))
```

Comment:
It would have been better if you could try not to realize the data, see https://scitools.org.uk/iris/docs/latest/userguide/real_and_lazy_data.html

If you need array functions to do this, use `from dask import array as da` instead of `import numpy as np`, see here for a description of dask array options: http://docs.dask.org/en/latest/array-api.html

Reply:
Sorry, I have merged this already.
Will do that in a separate PR. I think there other derivation function which may need to be changed as well.

Reply:
I think it's important that we try to address realizing the data in pull requests, because the amount of work it is to fix the non-lazy data everywhere will keep growing if we do not address it already when it is coming in.

Reply:
I have opened an issue #949: it looks like that 4 derived variables are affected at the moment.